ambry-tools/src/main/java/com.github.ambry/tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,975 @@
+        } finally {
+          futureResult.done(null, exceptionToReturn);
+          if (callback != null) {
+            callback.onCompletion(null, exceptionToReturn);
+          }
+        }
+      }
+    }
+
+    /**
+     * {@inheritDoc}
+     * Request to delete the blob from the server directly pertaining to the {@code blobId} passed in and verify that
+     * the subsequent get fails with {@link RouterErrorCode#BlobDeleted}
+     * @param blobId the blobId the of that blob that needs to be fetched
+     * @param expectedErrorCode the expected error code for the Delete call
+     * @param futureResult the {@link FutureResult} where the result/exception will be set on completion
+     * @param callback the {@link Callback} that will be called on completion
+     */
+    @Override
+    public void deleteBlobAndValidate(final String blobId, final RouterErrorCode expectedErrorCode,
+        final FutureResult futureResult, final Callback callback) {
+
+      try {
+        final Long startTimeGetBlob = SystemTime.getInstance().nanoseconds();
+        router.deleteBlob(blobId, new Callback<Void>() {
+          @Override
+          public void onCompletion(Void result, Exception exception) {
+            long latencyPerBlob = SystemTime.getInstance().nanoseconds() - startTimeGetBlob;
+            logger.trace(" Time taken to delete blob " + blobId + " in ms " + latencyPerBlob / SystemTime.NsPerMs);
+            final AtomicReference<Exception> exceptionToReturn = new AtomicReference<Exception>();
+            if (exception == null) {
+              logger.trace("Deletion of " + blobId + " succeeded. Issuing Get to confirm the deletion ");
+              FutureResult<GetBlobResult> getBlobResultFutureResult = new FutureResult<GetBlobResult>();
+              getBlobAndValidate(blobId, null, RouterErrorCode.BlobDeleted, getBlobResultFutureResult,
+                  new Callback<GetBlobResult>() {
+                    @Override
+                    public void onCompletion(GetBlobResult getBlobResult, Exception exception) {
+                      if (getBlobResult != null) {
+                        exceptionToReturn.set(new IllegalStateException(
+                            "Get of a deleted blob " + blobId + " should have failed with "
+                                + RouterErrorCode.BlobDeleted));
+                      } else {
+                        if (exception != null) {
+                          exceptionToReturn.set(exception);
+                        }
+                      }
+                    }
+                  });
+            } else {
+              if (exception instanceof RouterException) {
+                RouterException routerException = (RouterException) exception;
+                if (expectedErrorCode != null) {
+                  if (!expectedErrorCode.equals(routerException.getErrorCode())) {
+                    logger.trace("Error code mismatch. Expected " + expectedErrorCode + ", actual "
```

Comment:
I further think there is no need to log here. Maybe you could simply include the message in a new exception that wraps the original exception.